"""
Bioinformatics tools list,
format in ['tool name', 'replacements', 'Uppercase/Lowercase sensitive', 'tool url']
"""
BIO_TOOLS_MAP = [
    ['dss',                   'dss',                False,      'http://bioconductor.org/packages/release/bioc/html/DSS.html'],
    ['bwa',                   'bwa',                False,      'http://bio-bwa.sourceforge.net/'],
    ['maq',                   'maq',                False,      'http://maq.sourceforge.net/'],
    ['UCSC',                  'ucsctool',           True,       'https://genome.ucsc.edu/'],
    ['umls',                  'umls',               False,      'https://www.nlm.nih.gov/research/umls/'],
    ['pfam',                  'pfam',               False,      'https://pfam.xfam.org/'],
    ['affy',                  'affy',               False,      'https://bioconductor.org/packages/release/bioc/html/affy.html'],
    ['edmr',                  'edmr',               False,      'https://github.com/ShengLi/edmr'],
    ['gatk',                  'gatk',               False,      'https://software.broadinstitute.org/gatk/'],
    ['deseq',                 'deseq',              False,      'https://bioconductor.org/packages/release/bioc/html/DESeq.html'],
    ['rseqc',                 'rseqc',              False,      'https://bioconda.github.io/recipes/rseqc/README.html'],
    ['CREST',                 'cresttool',          True,       'https://www.stjuderesearch.org/site/lab/zhang'],
    ['tabix',                 'tabix',              False,      'http://www.htslib.org/doc/tabix.html'],
    ['Keras',                 'keras',              False,      'https://keras.io/'],
    ['gfold',                 'gfold',              False,      'https://zhanglab.tongji.edu.cn/softwares/GFOLD/index.html'],
    ['seqtk',                 'seqtk',              False,      'https://bioconda.github.io/recipes/seqtk/README.html'],
    ['plink',                 'plink',              False,      'http://zzz.bwh.harvard.edu/plink/'],
    ['mxnet',                 'mxnet',              False,      'https://mxnet.apache.org/'],
    ['caffe',                 'caffe',              False,      'http://caffe.berkeleyvision.org/'],
    ['edgeR',                 'edger',              False,      'https://bioconductor.org/packages/release/bioc/html/edgeR.html'],
    ['HTSeq',                 'htseq',              False,      'https://htseq.readthedocs.io/en/release_0.9.1/overview.html'],
    ['BLAST',                 'blast',              True,       'https://blast.ncbi.nlm.nih.gov/Blast.cgi'],
    ['Bpipe',                 'bpipe',              False,      'http://docs.bpipe.org/'],
    ['FastQC',                'fastqc',             False,      'https://www.bioinformatics.babraham.ac.uk/projects/fastqc/'],
    ['Ruffus',                'ruffus',             False,      'http://www.ruffus.org.uk/'],
    ['EMBOSS',                'embosstool',         False,      'http://www.bioinformatics.nl/emboss-explorer/'],
    ['Beagle',                'beagletool',         True,       'https://faculty.washington.edu/browning/beagle/beagle.html'],
    ['matlab',                'matlab',             False,      'https://www.mathworks.com/products/matlab.html'],
   # ['python',                'python',             False,      'https://www.python.org/'],
    ['cruzdb',                'cruzdb',             False,      'https://github.com/brentp/cruzdb'],
    ['snpeff',                'snpeff',             False,      'https://bioconda.github.io/recipes/snpeff/README.html'],
    ['rasmol',                'rasmol',             False,      'http://www.openrasmol.org/'],
    ['glasso',                'glasso',             False,      'https://cran.r-project.org/web/packages/glasso/'],
    ['Bowtie',                'bowtie',             False,      'http://bowtie-bio.sourceforge.net/index.shtml'],
    ['bionet',                'bionet',             False,      'http://bioconductor.org/packages/release/bioc/html/BioNet.html'],
    ['SHRiMP',                'shrimptool',         True,       'http://compbio.cs.toronto.edu/shrimp/'],
    ['bedops',                'bedops',             False,      'https://bedops.readthedocs.io/en/latest/'],
    ['tophat',                'tophat',             False,      'http://ccb.jhu.edu/software/tophat/index.shtml'],
    ['glmnet',                'glmnet',             False,      'https://cran.r-project.org/web/packages/glmnet/index.html'],
    ['samstat',               'samstat',            False,      'http://samstat.sourceforge.net/'],
    ['pegasus',               'pegasus',            False,      'https://pegasus.isi.edu/'],
    ['soapsnp',               'soapsnp',            False,      'http://soap.genomics.org.cn/soapsnp.html'],
    ['clustal',               'clustal',            False,      'http://www.clustal.org/'],
    ['mercury',               'mercury',            False,      'https://omictools.com/mercury-tool'],
    ['bismark',               'bismark',            False,      'https://www.bioinformatics.babraham.ac.uk/projects/bismark/'],
    ['bamUtil',               'bamutil',            False,      'https://genome.sph.umich.edu/wiki/BamUtil'],
    ['prtools',               'prtools',            False,      'http://prtools.org/'],
    ['sklearn',               'sklearn',            False,      'https://scikit-learn.org/stable/'],
    ['STAR',                  'startool',           True,       'https://github.com/alexdobin/STAR'],
    ['Graphviz',              'graphviz',           False,      'http://www.graphviz.org/'],
    ['ngsutils',              'ngsutils',           False,      'https://github.com/ngsutils/ngsutils'],
    ['methpipe',              'methpipe',           False,      'http://smithlabresearch.org/software/methpipe/'],
    ['seqtools',              'seqtools',           False,      'http://smithlabresearch.org/software/methpipe/'],
    ['igvtools',              'igvtools',           False,      'https://software.broadinstitute.org/software/igv/igvtools'],
    ['BAMStats',              'bamstats',           False,      'http://bamstats.sourceforge.net/'],
    ['clustalw2',             'clustalw',           False,      'https://www.ebi.ac.uk/Tools/msa/clustalw2/'],
    ['CellNOptR',             'cellnopt',           False,      'https://saezlab.github.io/CellNOptR/'],
    ['PROSPECT',              'prospecttool',       True,       'https://omictools.com/prospect-2-tool'],
    ['BiRewire',              'birewire',           False,      'http://bioconductor.org/packages/release/bioc/html/BiRewire.html'],
    ['cutadapt',              'cutadapt',           False,      'https://bioconda.github.io/recipes/cutadapt/README.html'],
    ['bamtools',              'bamtools',           False,      'https://github.com/pezmaster31/bamtools'],
    ['biopieces',             'biopieces',          False,      'http://maasha.github.io/biopieces/'],
   # ['c\+\+',                 'cplusplus',          False,      'https://en.cppreference.com/w/'],
    ['icommands',             'icommands',          False,      'https://wiki.cyverse.org/wiki/display/DS/Using+iCommands'],
    ['kentUtils',             'kentutils',          False,      'https://github.com/ENCODE-DCC/kentUtils'],
    ['TFBSTools',             'tfbstools',          False,      'https://bioconductor.org/packages/release/bioc/html/TFBSTools.html'],
    ['TrimGalore',            'trimgalore',         False,      'https://github.com/FelixKrueger/TrimGalore'],
    ['tensorflow',            'tensorflow',         False,      'https://www.tensorflow.org/'],
    ['cummeRbund',            'cummerbund',         False,      'https://bioconductor.org/packages/release/bioc/html/cummeRbund.html'],
    ['Snippy',                'snippytool',         True,       'https://github.com/tseemann/snippy'],
    ['SpliceTrap',            'splicetrap',         False,      'http://rulai.cshl.edu/splicetrap/'],
    ['Picard',                'picardtool',         True,       'https://github.com/broadinstitute/picard'],
    ['Glide',                 'glidetool',          True,       'https://omictools.com/glide-tool'],
    ['GenomeTools',           'genometools',        False,      'http://genometools.org/'],
    ['BreakDancer',           'breakdancer',        False,      'http://breakdancer.sourceforge.net/'],
    ['SOAPaligner',           'soapaligner',        False,      'http://soap.genomics.org.cn/soapaligner.html'],
    ['TopHat-Fusion',         'tophatfusion',       True,       'https://ccb.jhu.edu/software/tophat/fusion_index.shtml'],
    ['patternhunter',         'patternhunter',      False,      'https://omictools.com/patternhunter-tool'],
    ['bcl2fastq',             'bcl2fastq',          False,      'https://omictools.com/bcl2fastq-conversion-software-tool'],
    ['BiocGenerics',          'biocgenerics',       False,      'http://bioconductor.org/packages/3.8/bioc/html/BiocGenerics.html'],
    ['IRanges',               'iranges',            False,      'http://bioconductor.org/packages/3.8/bioc/html/IRanges.html'],
    ['BiQAnalyzer',           'biqanalyzer',        False,      'https://biq-analyzer.bioinf.mpi-inf.mpg.de/'],
    ['GeneGenie',             'genegenie',          False,      'http://g.gene-genie.appspot.com '],
    ['MoD Tools',             'modtools',           True,       'http://159.149.109.16/modtools/'],
    ['SVMHC',                 'svmhc',              False,      'http://www-abi.informatik.uni-tuebingen.de/Services/SVMHC/'],
    ['SciTools',              'scitools',           False,      'https://www.idtdna.com/pages/tools'],
    ['O-miner',               'ominer',             False,      'http://www.o-miner.org '],
    ['REPRO',                 'repro',              False,      'http://www.ibi.vu.nl/programs/reprowww/'],
    ['McPromoter',            'mcpromoter',         False,      'http://genes.mit.edu/McPromoter.html'],
    ['BiERapp',               'bierapp',            False,      'http://bierapp.babelomics.org/'],
    ['SOURCE',                'sourcetool',         True,       'http://source-search.princeton.edu/'],
    ['BioPortal',             'bioportal',          False,      'http://bioportal.bioontology.org/'],
    ['mGene',                 'mgene',              False,      'http://www.mgene.org/web'],
    ['Taverna',               'taverna',            False,      'https://taverna.incubator.apache.org/'],
    ['ProteinCCD',            'proteinccd',         False,      'http://xtal.nki.nl/ccd'],
    ['GenomeNet',             'genomenet',          False,      'https://www.genome.jp/'],
    ['YeTFaSCo',              'yetfasco',           False,      'http://yetfasco.ccbr.utoronto.ca/'],
    ['AnnotQTL',              'annotqtl',           False,      'http://annotqtl.genouest.org/'],
    ['GenePattern',           'genepattern',        False,      'http://software.broadinstitute.org/cancer/software/genepattern/'],
    ['PHIRE',                 'phire',              False,      'https://www.biw.kuleuven.be/logt/PHIRE.htm'],
    ['HaploReg',              'haploreg',           False,      'https://pubs.broadinstitute.org/mammals/haploreg/haploreg.php'],
    ['Satellog',              'satellog',           False,      'http://satellog.bcgsc.ca/'],
    ['PlantTFDB',             'plntfdb',            False,      'http://plntfdb.bio.uni-potsdam.de/v3.0/'],
    ['MODEST',                'modesttool',         True,       'http://modest.biosustain.dtu.dk/'],
    ['SoRT2',                 'sorttool',           False,      'http://genome.cs.nthu.edu.tw/SORT2/'],
    ['BioProspector',         'bioprospector',      False,      'http://robotics.stanford.edu/~xsliu/BioProspector/'],
    ['PubCrawler',            'pubcrawler',         False,      'http://pubcrawler.gen.tcd.ie/'],
    ['DINAMelt',              'dinamelt',           False,      'http://unafold.rna.albany.edu/?q=DINAMelt'],
    ['Phyloscan',             'phyloscan',          False,      'http://bayesweb.wadsworth.org/phyloscan/'],
    ['MIRU-VNTRplus',         'miru',               False,      'https://www.miru-vntrplus.org/MIRU/index.faces'],
    ['RepeatMasker',          'repeatmasker',       False,      'http://www.repeatmasker.org/'],
    ['STAMP',                 'stamptool',          True,       ''],
    ['EMBL-EBI',              'ebi',                False,      'https://www.ebi.ac.uk/about/our-impact'],
    ['VarMod',                'varmod',             False,      'http://www.wasslab.org/varmod/'],
    ['CHOPCHOP',              'chopchop',           False,      'http://chopchop.cbu.uib.no/'],
    ['Primer3Plus',           'primerplus',         False,      'http://www.bioinformatics.nl/cgi-bin/primer3plus/primer3plus.cgi'],
    ['CARGO',                 'cargotool',          True,       'http://cargo2.bioinfo.cnio.es/docs/aboutCargo.html'],
    ['GLIMMER',               'glimmertool',        False,      'http://ccb.jhu.edu/software/glimmer/index.shtml'],
    ['GeneMark',              'genemark',           False,      'http://opal.biology.gatech.edu/GeneMark/'],
    ['VISTA',                 'vistatool',          False,      'http://genome.lbl.gov/vista/index.shtml'],
    ['cipres',                'cipres',             False,      'http://www.phylo.org/'],
    ['RNAmmer',               'rnammer',            False,      'http://www.cbs.dtu.dk/services/RNAmmer/'],
    ['PHAST',                 'phast',              False,      'http://phast.wishartlab.com/'],
    ['BoxShade',              'boxshade',           False,      'http://www.ch.embnet.org/software/BOX_form.html'],
    ['inGAP',                 'ingap',              False,      'http://ingap.sourceforge.net/'],
    ['fpc',                   'fpc',                False,      'http://www.agcol.arizona.edu/software/fpc/'],
    ['Orphelia',              'orphelia',           False,      'http://orphelia.gobics.de/'],
    ['PeakSeq',               'peakseq',            False,      'http://info.gersteinlab.org/PeakSeq'],
    ['PANTHER',               'pantherdb',          True,       'http://www.pantherdb.org/tools/'],
    ['NJplot',                'njplot',             False,      'http://doua.prabi.fr/software/njplot'],
    ['IMP',                   'imp',                False,      'http://imp.princeton.edu/'],
    ['islandpath',            'islandpath',         False,      'http://www.pathogenomics.sfu.ca/islandpath/'],
    ['fidea',                 'fidea',              False,      'http://circe.med.uniroma1.it/fidea/'],
    ['corgen',                'corgen',             False,      'http://corgen.molgen.mpg.de/'],
    ['Clone DB',              'clonedb',            True,       'https://www.ncbi.nlm.nih.gov/clone'],
    ['FirstEF',               'firstef',            False,      'http://rulai.cshl.org/tools/FirstEF/'],
    ['miRmap',                'mirmap',             False,      'https://mirmap.ezlab.org/'],
    ['rfam',                  'rfam',               False,      'https://rfam.sanger.ac.uk/'],
    ['itbtools',              'itbtools',           False,      'http://itbtools.ba.itb.cnr.it/patsearch'],
    ['SortMeRNA',             'sortmerna',          False,      'http://bioinfo.lifl.fr/RNA/sortmerna/'],
    ['IntaRNA',               'intarna',            False,      'http://rna.informatik.uni-freiburg.de/IntaRNA/Input.jsp'],
    ['pknotsrg',              'pknotsrg',           False,      'https://bibiserv.cebitec.uni-bielefeld.de/pknotsrg/'],
    ['snoscan',               'snoscan',            False,      'http://lowelab.ucsc.edu/snoscan/'],
    ['taxmanwww',             'taxmanwww',          False,      'http://www.ibi.vu.nl/programs/taxmanwww/'],
    ['Sfold',                 'sflod',              False,      'http://sfold.wadsworth.org/cgi-bin/index.pl'],
    ['ERPIN',                 'erpin',              False,      'http://rna.igmors.u-psud.fr/Software/erpin.php'],
    ['PatMatch',              'patmatch',           False,      'https://www.arabidopsis.org/cgi-bin/patmatch/nph-patmatch.pl'],
    ['alexa_seq',             'alexaseq',           False,      'http://www.alexaplatform.org/alexa_seq/'],
    ['RNALOSS',               'rnaloss',            False,      'http://bioinformatics.bc.edu/clotelab/RNALOSS/'],
    ['corrna',                'corrna',             False,      'http://corrna.cs.mcgill.ca/'],
    ['DARIO',                 'dario',              False,      'http://dario.bioinf.uni-leipzig.de/'],
    ['comir',                 'comir',              False,      'http://www.benoslab.pitt.edu/comir/'],
    ['Foldalign',             'foldalign',          False,      'https://rth.dk/resources/foldalign/'],
    ['rnapdbee',              'rnapdbee',           False,      'http://rnapdbee.cs.put.poznan.pl/'],
    ['toppmir',               'toppmir',            False,      'https://toppmir.cchmc.org/'],
    ['STarMir',               'starmir',            False,      'http://sfold.wadsworth.org/cgi-bin/starmirtest2.pl'],
    ['e-RNA',                 'erna',               False,      'http://www.e-rna.org/'],
    ['sivirus',               'sivirus',            False,      'http://sivirus.rnai.jp/'],
    ['togotable',             'togotable',          False,      'http://togotable.dbcls.jp/'],
    ['NAPP',                  'napp',               False,      'http://napp.u-psud.fr/'],
    ['rbpmap',                'rbpmap',             False,      'http://rbpmap.technion.ac.il/'],
    ['Magnolia',              'magnoliatool',       True,       'http://bioinfo.lifl.fr/magnolia/magnolia.php'],
    ['OligoWalk',             'oligowalk',          False,      'http://rna.urmc.rochester.edu/cgi-bin/server_exe/oligowalk/oligowalk_form.cgi'],
    ['rnasoft',               'rnasoft',            False,      'http://www.rnasoft.ca/'],
    ['nobai',                 'nobai',              False,      'http://manet.illinois.edu/nobai/nobai.php'],
    ['cognac',                'cognac',             False,      'http://27.126.156.175/cognac/'],
    ['molmovdb',              'molmovdb',           False,      'http://molmovdb.org/'],
    ['siret',                 'siret',              False,      'http://siret.ms.mff.cuni.cz/'],
    ['molprobity',            'molprobity',         False,      'http://molprobity.biochem.duke.edu/'],
    ['WebFR3D',               'webfrthreed',        False,      'https://www.bgsu.edu/research/rna/web-applications/webfr3d.html'],
    ['snoGPS',                'snogps',             False,      'http://lowelab.ucsc.edu/snoGPS/'],
    ['cylofold',              'cylofold',           False,      'https://cylofold.ncifcrf.gov/'],
    ['SAVoR',                 'savortool',          True,       'https://savor.lisanwanglab.org/'],
    ['SCeQTL',                'sceqtl',             False,      'https://github.com/XuegongLab/SCeQTL/'],
    ['Gencore',               'gencore',            True,       'https://github.com/OpenGene/gencore/'],
    ['ZINBayes',              'zinbayes',           False,      'https://github.com/pedrofale/zinbayes'],
    ['m-pCMF',                'mpcmf',              False,      'https://github.com/pedrofale/mpCMF'],
    ['SPRING',                'springtool',         True,       'https://github.com/shubhamchandak94/SPRING'],
    ['Alfred',                'alfred',             True,       'https://tobiasrausch.com/alfred/'],
    ['SCLUST',                'sclust',             False,      'https://omictools.com/sclust-tool'],
    ['PopCorn',               'popcorn',            True,       'https://omictools.com/popcorn-tool'],
    ['RISC',                  'risctool',           True,       'https://omictools.com/risc-tool'],
    ['FastqPuri',             'fastqpuri',          False,      'https://github.com/jengelmann/FastqPuri'],
    ['VCFtools',              'vcftools',           False,      'https://vcftools.github.io/index.html'],
    ['Samtools',              'samtools',           False,      'http://www.htslib.org/'],
    ['BEDTools',              'bedtools',           False,      'https://github.com/arq5x/bedtools2'],
    ['Cufflinks',             'cufflinks',          False,      'http://cole-trapnell-lab.github.io/cufflinks/'],
    ['Mikado',                'mikado',             False,      'https://github.com/lucventurini/Mikado'],
    ['fast5_scripts',         'fastscripts',        False,      'https://github.com/nick297/fast5_scripts'],
    ['WagnerEtAl2017',        'wagneretal',         False,      'https://omictools.com/wagneretal2017-tool'],
    ['BIO_CYCLE',             'biocycle',           False,      'https://omictools.com/bio-cycle-tool'],
    ['circadiomics',          'circadiomics',       False,      'http://circadiomics.igb.uci.edu/'],
    ['MUSCLE',                'muscletool',         True,       'https://www.drive5.com/muscle/'],
    ['HapBlock',              'hapblock',           False,      'https://www.soph.uab.edu/sites/edu.ssg/files/People/KZhang/HapBlock/hapblock-index.html'],
    ['GeneSeeker',            'geneseeker',         False,      'http://www.cmbi.ru.nl/GeneSeeker/'],
    ['COBRA',                 'cobratool',          True,       'https://opencobra.github.io/cobratoolbox/stable/'],
    ['gCLUTO',                'gcluto',             False,      'http://glaros.dtc.umn.edu/gkhome/cluto/gcluto/overview'],
    ['Pafig',                 'pafig',              False,      'http://www.mobioinfor.cn/pafig/'],
    ['CisModule',             'cismodule',          False,      'http://www.stat.ucla.edu/~zhou/CisModule/'],
    ['ModuleDigger',          'moduledigger',       False,      'http://bioinformatics.intec.ugent.be/kmarchal/Supplementary_Information_Sun_2009_ModuleDigger/Index.html'],
    ['MGR',                   'mgr',                False,      'https://omictools.com/mgr-tool'],
    ['ComPhy',                'comphy',             False,      'http://digbio.missouri.edu/ComPhy/'],
    ['phylip',                'phylip',             False,      'http://evolution.genetics.washington.edu/phylip.html'],
    ['Haploview',             'haploview',          False,      'https://www.broadinstitute.org/haploview/haploview'],
    ['mgrep',                 'mgrep',              False,      ''],
    ['GenotypeColour',        'genotypecolour',     False,      ''],
    ['Medical Text Indexer',  'mti',                False,      'https://ii.nlm.nih.gov/MTI/'],
    ['Cyber-T',               'cybert',             False,      'https://omictools.com/cyber-t-tool'],
    ['BioMart',               'biomart',            False,      'http://www.biomart.org/'],
    ['MimMiner',              'mimminer',           False,      'https://omictools.com/mimminer-tool'],
    ['Autodock',              'autodock',           False,      'http://autodock.scripps.edu/'],
    ['Gromacs',               'gromacs',            False,      'http://www.gromacs.org/'],
    ['mfold',                 'mfold',              False,      ''],
    ['cplex',                 'cplex',              False,      ''],
    ['psipred',               'psipred',            False,      'http://bioinf.cs.ucl.ac.uk/psipred/'],
    ['MEGAN',                 'megantool',          False,      ''],
    ['PyTorch',               'pytorch',            False,      'https://pytorch.org/'],
    ['Theano',                'theano',             False,      'http://www.deeplearning.net/software/theano/'],
]


def check_validity():
    d = {}
    for i in BIO_TOOLS_MAP:
        # check length
        if len(i) != 4:
            print "Info not compeleted: " + str(i)

        # check repeated
        if i[1] in d:
            print "Repeated tool %s found!" % str(i[1])
        else:
            d[i[1]] = 1

        # check lower case
        if i[1].lower() != i[1]:
            print "%s is not lowercase" % str(i[1])


if __name__ == "__main__":
    check_validity()
